include/circt/Dialect/FIRRTL/FIRRTLTypes.td

```diff
@@ -236,4 +236,18 @@ def DoubleType : FIRRTLDialectTypeHelper<"DoubleType", "double type">,
 def PathType : FIRRTLDialectTypeHelper<"PathType", "path type">,
   BuildableType<"::circt::firrtl::PathType::get($_builder.getContext())">;
 
+//===----------------------------------------------------------------------===//
+// Format String Types
+//===----------------------------------------------------------------------===//
+
+def FStringTimeType :
+  FIRRTLDialectTypeHelper<"FStringTimeType",
+    "a format string type that represents the current simulation time">,
+  BuildableType<"::circt::firrtl::FStringTimeType::get($_builder.getContext())">;
+
+def PrintfOperandType : FIRRTLDialectType<
+  CPred<"type_isa<FIRRTLBaseType, FStringTimeType>($_self)">,
+    "a printf operand type (a FIRRTL base type or a format string type)",
+    "::circt::firrtl::FIRRTLType">;
+
```

Comment:
I'm wondering if we need to distinguish between the exact nature of the format string snippets at all. We might be able to get away with just a generic `FStringType` which all interpolator ops produce. Something like this:

```mlir
%0 = firrtl.fstring.time ... : !firrtl.fstring
%1 = firrtl.fstring.literal ": got wrong value " : !firrtl.fstring
%2 = firrtl.fstring.int hex, width 16, padzero, ... : !firrtl.fstring
%3 = firrtl.fstring.concat %0, %1, %2 : !firrtl.fstring
firrtl.printf %clock, %c1_ui1, %3
```

WDYT?

Reply:
I went back and forth on this. I originally coded it this way and then changed it for reasons that aren't entirely clear. I think I was originally wanting to have some information about the type so that the operand could know what it is without looking at the op. Though, that seems bad.

Let me rework this in that direction. I agree that you eventually just want a single `fstring` type that other ops produce, like concat, and having this blow up with many different per-op types is wrong.

Reply:
Changed to this in: 41e10163f